compiler/luci/pass/src/FuseSliceWithTConvPass.cpp

```diff
@@ -0,0 +1,287 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FuseSliceWithTConvPass.h"
+
+#include <luci/Service/CircleNodeClone.h>
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+namespace
+{
+
+#define RETURN_FALSE_UNLESS(cond) \
+  if (not(cond))                  \
+    return false;
+
+inline int32_t compute_input_size(luci::Padding padding, int32_t image_size, int32_t filter_size,
+                                  int32_t stride)
+{
+  switch (padding)
+  {
+    case luci::Padding::SAME:
+      return (image_size + stride - 1) / stride;
+    case luci::Padding::VALID:
+      return (image_size + stride - filter_size) / stride;
+    default:
+      assert(false);
+      return 0;
+  }
+}
+
+inline int32_t extract_pad_value(int32_t stride, int32_t in_size, int32_t filter_size,
+                                 int32_t out_size)
+{
+  const int32_t padding = ((in_size - 1) * stride + filter_size - out_size) / 2;
+  return padding > 0 ? padding : 0;
+}
+
+inline uint32_t cal_offset(const luci::CircleConst *node, const uint32_t *indices)
+{
+  return indices[0] * node->dim(1).value() * node->dim(2).value() * node->dim(3).value() +
+         indices[1] * node->dim(2).value() * node->dim(3).value() +
+         indices[2] * node->dim(3).value() + indices[3];
+}
+
+luci::Padding compute_padding(const luci::CircleTransposeConv *tconv, int32_t out_height,
+                              int32_t out_width, int32_t pad_top, int32_t pad_left)
+{
+  auto const filter = dynamic_cast<luci::CircleConst *>(tconv->filter());
+  if (!filter)
+    return luci::Padding::UNDEFINED;
+
+  auto tconv_shape = dynamic_cast<luci::CircleConst *>(tconv->inputSizes());
+  if (!tconv_shape)
+    return luci::Padding::UNDEFINED;
+
+  luci::Padding padding = luci::Padding::UNDEFINED;
+  std::initializer_list<luci::Padding> paddings_to_check = {luci::Padding::VALID,
+                                                            luci::Padding::SAME};
+
+  auto const filter_height = filter->dim(1).value();
+  auto const filter_width = filter->dim(2).value();
+  auto const stride_height = tconv->stride()->h();
+  auto const stride_width = tconv->stride()->w();
+
+  for (auto padding_to_check : paddings_to_check)
+  {
+    auto const in_height =
+      compute_input_size(padding_to_check, out_height, filter_height, stride_height);
+    auto const pad_top_virtual =
+      extract_pad_value(stride_height, in_height, filter_height, out_height);
+    if (pad_top_virtual != pad_top)
+      continue;
+
+    auto const in_width =
+      compute_input_size(padding_to_check, out_width, filter_width, stride_width);
+    auto const pad_left_virtual =
+      extract_pad_value(stride_width, in_width, filter_width, out_width);
+    if (pad_left_virtual == pad_left)
+    {
+      padding = padding_to_check; // correct padding is found
```

Comment:
Q) Is it okay not to check bottom/right padding?

Reply:
> Q) Is it okay not to check bottom/right padding?

As far as i know - yes. It's okay.  Bottom/right paddings are just for correcting size of resulting TConv dimension.

Reply:
Can you add a test for this case in draft, i.e., top padding != bottom padding?

Reply:
ok

Reply:
@jinevening 
To clarify:
https://github.com/Samsung/ONE/blob/b284f4f63ca5049385a7dd8c47a8e535b86617e9/compiler/luci-interpreter/src/kernels/TransposeConv.cpp#L73-L76

TConv kernel doesn't use pad_bottom or pad_right, they are used implicitly via spatial size reduction.
So this PR doesn't use bottom/right pads.
The only effort made in this PR is to make sure values for top/left padding, (which are the only values used after deciphering `VALID/SAME+kernel+shape=>(top/left pads)`) will be the same as those used in `luci-interpreter'. 
Actually current test in draft (it's yet `TConvSlice_000/test.recipe`) uses (pad_top = pad_left = 0)  while (pad_bottom = pad_right = 3).

Reply:
You've already added tests. Thanks.

For `Net_TConv_Slice_000`,
- Before
  - [1, 8, 8, 32] ->  TConv [1, 18, 18, 32] (valid) -> Slice [1, 15, 15, 32]
  - Slice Op slices like `pad_top` = 0, `pad_bottom` = 3.
- After
  - [1, 8, 8, 32] ->  TConv [1, 15, 15, 32] (same)
  - Padding would be `pad_top` = 1, `pad_bottom` = 2 to make them balanced as much as possible.

The outputs would be different, but the test passed.. Could you help my understanding? What am I missing here?

Reply:
> Could you help my understanding?

@jinevening 
As far as i understand,
the only thing that matters for interpreter is `_padding_height/_padding_left` at https://github.com/Samsung/ONE/blob/b284f4f63ca5049385a7dd8c47a8e535b86617e9/compiler/luci-interpreter/src/kernels/TransposeConv.cpp#L73-L76.

> * [1, 8, 8, 32] ->  TConv [1, 15, 15, 32] (same)

To my knowledge padding of resulting TConv would be `valid`, please see the results of running pass in attachment.
[Net_TConv_Slice_000..zip](https://github.com/Samsung/ONE/files/13050547/Net_TConv_Slice_000.zip)
so resulting pads for `valid` tconv would be (pad_top = 0, pad_bottom = 3), the same as in `slice`.

Please note that the code in PR tries to produce such values for tconv's `padding`, which would produce the same values for `_padding_height/_padding_left`  in interpreter.

Reply:
@jinevening 
Just a note:
I believe that seamless fusion of `Slice` with `TConv` would be possible if there was something like `PaddedTConv` with an option to explicitly set pad numerical values.

Reply:
> Thanks for explanation.

Glad to be helpful.